### PR TITLE
Rule config lookup improvements & config warnings

### PR DESCRIPTION
**Pull request by alanmcruickshank**
This PR is part of #4030 and stems from a conversation between @alanmcruickshank and @pwildenhain .

It sets the groundwork for being able to configure rules using the `name` in the config rather than the `code`. This does all the logic changes, but doesn't actually flip the switch yet (see `BaseRule.get_config_ref()` still returns `self.code` for now).

When the method is switched to return `self.name` for the rules which implement it, rules will be default look for their config at `sqlfluff:rules:my_rule_name`, which is much easier to interpret and understand when reading a config file. To avoid the ambiguity of having multiple potential config settings, where a rule does define a name - it can _only be configured by the name_. If a rule does not define a name, _it can only be configured by the code_. This means there is no ambiguity, but still allows backward compatibility for plugins which don't yet define a `name`.

This PR moves the `code` and `description` extraction logic into the metaclass introduced in #4415, this should have a minor performance gain but mostly a simpler ability to extract the `code` and `description` before the instantiation of the rule.

This PR _also_ does explicit checking against the provided configuration. If any rule config sections are found which _don't_ correspond to a valid rule reference, a warning will be shown. This has several upsides:
- on releasing 2.0.0, this is a convenient way to warn people they need to update their config files and provides hints on how to do so.
- if config files contain errors in the section headings, they will now be warned to the user rather than silently ignored.
- internally within the project we can be more confident that there aren't errors in our config because they will be explicitly raised.

I chose to _warn_ rather than error so that people can continue to still use the tool.

Additionally, this solves two other small issues:
1. As brought up in [this comment](https://github.com/sqlfluff/sqlfluff/pull/4468#discussion_r1126445313), I've added some validation for rule names, enforcing _lowercase_ and _snake_case_
2. I've also closed a small loophole with group inheritance where groups for `CPO2` were not showing in the docs.

My intent is to stage a second PR to actually do the swap of the return value of `get_config_ref`, because that PR will be very large as it will require updating all of the configuration files in the project.

### CODE REVIEW

**Comment by alanmcruickshank** [comment]
hmmmm - needs more tests. Will do that before putting this up for review again.

File: src/sqlfluff/core/rules/base.py [hunk 1]
```diff
@@ -499,6 +499,7 @@ class RuleMetaclass(type):
         flags=re.MULTILINE,
     )
     _valid_classname_regex = regex.compile(r"Rule_?([A-Z]{1}[a-zA-Z]+)?_([A-Z0-9]{4})")
+    _valid_rule_name_regex = regex.compile(r"[a-z\.\_]+")
```
Comment (WittierDinosaur) [inline src/sqlfluff/core/rules/base.py:502]: thoughts on `a-z[a-z\.\_]+`?

Comment (alanmcruickshank) [reply src/sqlfluff/core/rules/base.py:502]: yeah, that would be better... I'll update it.

File: src/sqlfluff/core/rules/base.py [hunk 2]
```diff
@@ -614,6 +684,12 @@ class BaseRule(metaclass=RuleMetaclass):
     # referred to this rule.
     aliases: Tuple[str, ...] = ()
 
+    # NOTE: code and description are provided here as hints, but should not
+    # be set directly. They are set automatically by the metaclass based on
+    # the class _name_ when defined.
+    code: str = ""
```
Comment (WittierDinosaur) [inline src/sqlfluff/core/rules/base.py:690]: for hints, do you need to set a default value? Can't you simply
```
code: str
```

Comment (alanmcruickshank) [reply src/sqlfluff/core/rules/base.py:690]: Good question - I'll try it 🤷

Comment (alanmcruickshank) [reply src/sqlfluff/core/rules/base.py]: looks like that works. Change coming through now...

**Review by WittierDinosaur — APPROVED** [review]
LGTM

**Pull request merged**